pkg/tfshim/shim.go

```diff
@@ -207,3 +208,73 @@ type Provider interface {
 	NewResourceConfig(object map[string]interface{}) ResourceConfig
 	IsSet(v interface{}) ([]interface{}, bool)
 }
+
+type ProviderWithContext interface {
```

Comment:
I'm a little fuzzy on the goal of `ProviderWithContext`. This is intended as a backwards compatibility bridge within the repo?

Is the desired end state that everybody implements `ProviderWithContext` and then `Provider` methods are never called?

Would it be more correct to change the definition of `Provider` to have methods that take `context.Context` as their first arguments?

Reply:
Yea this is a very mechanical refactor that avoids making the change breaking. 

> Would it be more correct to change the definition of Provider to have methods that take context.Context as their first arguments?

That'd be our desired end state. I'm curious if we can just take that change without breaking any dependent code (that is the shim.Provider is fully internal interface although it's exported). That might be the case, but I'm not 100% sure. Doing so would also fix context propagation issues for SDKv1 providers if any (which seemingly is still in production). Though I don't know if there'd be any end-user benefit.

This should probably be a follow-up change though. I've avoided going this way in this PR to avoid some additional yak shaving necessary to get SDKv2, protov5 (unused in prod but hanging around in the repo), and test providers line up with the change.